### ISSUE TITLE
[2018.3][Centos 7][py2] integration.shell.test_call.CallTest.test_issue_14979_output_file_permissions

### ISSUE THREAD

**Issue by gtmanfred** [saltstack/salt-jenkins#1000]
https://jenkinsci.saltstack.com/job/2018.3.1/view/Tornado/job/salt-centos-7-py2-tornado/16/testReport/integration.shell.test_call/CallTest/test_issue_14979_output_file_permissions/
https://jenkinsci.saltstack.com/job/2018.3.1/job/salt-centos-7-py2/30/testReport/integration.shell.test_call/CallTest/test_issue_14979_output_file_permissions/

This is failing on regular and tornado tests.

but not py3

**Comment by rallytime** [comment]
@terminalmage This is failing on the branch still, and the following output was given:

```
Error Message
Failed to generate output file, see log for details
Stacktrace
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/integration/shell/test_call.py", line 433, in test_issue_14979_output_file_permissions
    'Failed to generate output file, see log for details')
AssertionError: Failed to generate output file, see log for details
		
Standard Output
	
Standard Error
Traceback (most recent call last):
  File "/usr/lib64/python2.7/logging/__init__.py", line 851, in emit
    msg = self.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 724, in format
    return fmt.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 464, in format
    record.message = record.getMessage()
  File "/usr/lib64/python2.7/logging/__init__.py", line 328, in getMessage
    msg = msg % self.args
TypeError: %d format: a number is required, not NoneType
Logged from file test_call.py, line 430
Traceback (most recent call last):
  File "/usr/lib64/python2.7/logging/__init__.py", line 851, in emit
    msg = self.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 724, in format
    return fmt.format(record)
  File "/usr/lib64/python2.7/logging/__init__.py", line 464, in format
    record.message = record.getMessage()
  File "/usr/lib64/python2.7/logging/__init__.py", line 328, in getMessage
    msg = msg % self.args
TypeError: %d format: a number is required, not NoneType
Logged from file test_call.py, line 430
```

This is on a CentOS 7 Py 2 test: https://jenkinsci.saltstack.com/job/2018.3/job/salt-centos-7-py2/lastCompletedBuild/testReport/

**Comment by terminalmage** [comment]
That was a bug in my logging, I fixed that and this was the result:

```
22:20:32,152 [integration.shell.test_call                :430 ][ERROR   ] run_script failed to generate output file:
return code: None
stdout:
['Process took more than 15 seconds to complete. Process Killed!']

stderr
['Process killed, unable to catch stderr output']
```

So, somehow it is taking 15 seconds to run a `salt-call -g`. It's still unclear what is causing this though since we don't have access to the logging for the salt-call, given that it was spawned in a subprocess and is not connected to the test suite's logging.

The next step is to add an `-l debug` to the output. I'm going to open another PR to modify this test to add `-l debug` to the salt-call command, and I'm also moving the logging I did in the test into `run_script` and making some DRY changes there to untangle the web of if statements that it currently uses to return output and return codes from the script that was run.

**Comment by terminalmage** [comment]
OK, so it looks like when we kill the process due to timeout, we don't get any of its output. So my recent changes didn't produce any new information in the log.

**Comment by terminalmage** [comment]
OK, https://github.com/saltstack/salt/pull/47976 will add the output from the salt-call to what is logged when we timeout.

**Comment by terminalmage** [comment]
Here's the logging. It's possible that one of the grains is timing out. Unfortunately, we have no timestamps, so I am going to turn the logging to `trace` level for this salt-call and then several branch builds run and see if it hangs on the same grains function each run.

```
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion
[DEBUG   ] Including configuration from '/tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf
[INFO    ] Processing `log_handlers.runtests_log_handler`
[DEBUG   ] Adding the 'runtests_log_handler' provided logging handler: '<salt.log.handlers.QueueHandler object at 0x7fa1158a6b50>'
[DEBUG   ] Configuration file path: /tmp/salt-tests-tmpdir/config/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion
[DEBUG   ] Including configuration from '/tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf
[DEBUG   ] pyVmomi not loaded: Incompatible versions of Python. See Issue #29537.
[DEBUG   ] Loading static grains from /tmp/salt-tests-tmpdir/config/grains
[DEBUG   ] Connecting to master. Attempt 1 of 5
[DEBUG   ] Master URI: tcp://127.0.0.1:64506
[DEBUG   ] Popen(['git', 'version'], cwd=/tmp/kitchen/testing, universal_newlines=False, shell=None)
[DEBUG   ] Popen(['git', 'version'], cwd=/tmp/kitchen/testing, universal_newlines=False, shell=None)
[DEBUG   ] Initializing new AsyncAuth for (u'/tmp/salt-tests-tmpdir/rootdir/pki', u'minion', u'tcp://127.0.0.1:64506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (2159)
[DEBUG   ] Setting zmq_reconnect_ivl to '2159ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'/tmp/salt-tests-tmpdir/rootdir/pki', u'minion', u'tcp://127.0.0.1:64506', 'clear')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://127.0.0.1:64506
[DEBUG   ] Trying to connect to: tcp://127.0.0.1:64506
[DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] salt.crypt._get_key_with_evict: Loading private key
[DEBUG   ] Loaded minion key: /tmp/salt-tests-tmpdir/rootdir/pki/minion.pem
[DEBUG   ] salt.crypt.get_rsa_pub_key: Loading public key
[DEBUG   ] Connecting the Minion to the Master publish port, using the URI: tcp://127.0.0.1:64505
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] Loaded minion key: /tmp/salt-tests-tmpdir/rootdir/pki/minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for (u'/tmp/salt-tests-tmpdir/rootdir/pki', u'minion', u'tcp://127.0.0.1:64506', u'aes')
[DEBUG   ] Initializing new AsyncAuth for (u'/tmp/salt-tests-tmpdir/rootdir/pki', u'minion', u'tcp://127.0.0.1:64506')
[DEBUG   ] Connecting the Minion to the Master URI (for the return server): tcp://127.0.0.1:64506
[DEBUG   ] Trying to connect to: tcp://127.0.0.1:64506
[DEBUG   ] salt.crypt.get_rsa_key: Loading private key
[DEBUG   ] Loaded minion key: /tmp/salt-tests-tmpdir/rootdir/pki/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion
[DEBUG   ] Including configuration from '/tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /tmp/salt-tests-tmpdir/config/minion.d/_schedule.conf
```

**Comment by terminalmage** [comment]
The first run with the trace logging shows that it timed out while in the `os_data()` func in `salt/grains/core.py`. Another build is running now, and if it stops in the same place it will suggest that something in the `os_data()` func is timing out, and I'll know where to add further trace logging to zero in on it.

```
[TRACE   ] Loading core.append_domain grain
[TRACE   ] Loading core.default_gateway grain
[TRACE   ] Loading core.dns grain
[TRACE   ] Loading core.get_machine_id grain
[TRACE   ] Loading core.get_master grain
[TRACE   ] Loading core.get_server_id grain
[TRACE   ] Loading core.hostname grain
[TRACE   ] Loading core.hwaddr_interfaces grain
[TRACE   ] Loading core.id_ grain
[TRACE   ] Loading core.ip4_interfaces grain
[TRACE   ] Loading core.ip6_interfaces grain
[TRACE   ] Loading core.ip_fqdn grain
[TRACE   ] Loading core.ip_interfaces grain
[TRACE   ] Loading core.linux_distribution grain
[TRACE   ] Loading core.locale_info grain
[TRACE   ] Loading core.os_data grain
```

**Comment by terminalmage** [comment]
After getting the trace logging into the salt-call, I was not able to find anything that took unusually long. What I did notice however is that `salt-call -g` was compiling grains twice! After some digging I found that the `salt.cli.caller.BaseCaller`'s `print_grains()` function was compiling the grains before printing. However, this class also creates a `salt.minion.SMinion` instance in its dunder init, and the _`SMinion`_'s dunder init is _already_ compiling grains in _its_ dunder init. I opened https://github.com/saltstack/salt/pull/48286 which I found reduced the time it took to run `salt-call -g` to almost half.
